plugins/logs/mask.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2020 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package logs
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/open-policy-agent/opa/util"
+)
+
+type maskOP string
+
+const (
+	maskOPRemove maskOP = "remove"
+	maskOPUpsert maskOP = "upsert"
+)
+
+type maskRule struct {
+	OP           maskOP      `json:"op"`
+	Path         string      `json:"path"`
+	Value        interface{} `json:"value"`
+	escapedParts []string
+}
+
+func (r maskRule) String() string {
+	return "/" + strings.Join(r.escapedParts, "/")
+}
+
+func newMaskRule(op maskOP, path string, val interface{}) (maskRule, error) {
+	rule := maskRule{}
+
+	if len(path) == 0 {
+		return rule, fmt.Errorf("mask must be non-empty")
+	} else if path[0] != '/' {
+		return rule, fmt.Errorf("mask must be slash-prefixed")
+	}
+
+	parts := strings.Split(path[1:], "/")
+
+	if parts[0] != "input" && parts[0] != "result" {
+		return rule, fmt.Errorf("mask prefix not allowed")
+	}
+
+	escapedParts := make([]string, len(parts))
+	for i := range parts {
+		_, err := url.QueryUnescape(parts[i])
+		if err != nil {
+			return rule, err
+		}
+
+		escapedParts[i] = url.QueryEscape(parts[i])
+	}
+
+	rule = maskRule{
+		OP:           op,
+		Path:         path,
+		Value:        val,
+		escapedParts: escapedParts,
+	}
+
+	return rule, nil
+}
+
+func (r maskRule) Mask(event *EventV1) error {
+
+	if len(r.escapedParts) == 1 {
+		switch p := r.escapedParts[0]; p {
+		case "input":
```

Comment:
We can probably move the common code below into a function.

Reply:
Refactored the Mask function.